src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,156 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            createFirstNode(value);
+        } else {
+            Node<T> nextNode = new Node<>(tail, value, null);
+            nextNode.prev.next = nextNode;
+            tail = nextNode;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        isValidIndexAdd(index);
+        if (index == size()) {
+            add(value);
+        } else if (index == 0) {
+            addToTheBegin(value);
+        } else {
+            addToTheMiddle(value, index);
+        }
+    }
+
+    private void addToTheBegin(T value) {
+        Node<T> newNode = new Node<>(null, value, head);
+        newNode.next.prev = newNode;
+        head = newNode;
+    }
+
+    private void addToTheMiddle(T value, int index) {
+        Node<T> current = head;
+        int count = 0;
+        while (count++ != index) {
+            current = current.next;
+        }
+        Node<T> newNode = new Node<>(current.prev, value, current);
+        newNode.prev.next = newNode;
+        newNode.next.prev = newNode;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null || list.isEmpty()) {
+            return;
+        }
+        for (T node : list) {
+            add(node);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = getNode(index);
+        T oldValue = node.element;
+        node.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> current = getNode(index);
+        T oldValue = current.element;
+        if (size() == 1) {
+            head = null;
+            tail = null;
+        } else if (index == 0) {
+            current.next.prev = null;
+            head = current.next;
+        } else if (index == (size() - 1)) {
+            current.prev.next = null;
+            tail = current.prev;
+        } else {
```

Comment:
too much if-else-else if. we can simplify this construction getting previous and next of current node and check them for null